### PR TITLE
Fix json parsing bug with plugins that don't provide args

### LINKED ISSUE #5484: JSON Plugin config in 0.9-rc1 non functional
Not sure if this is a question or a bug 🙂 

### Nomad version
`Nomad v0.9.0-rc1 (7c00ab4f3f37cfd1e258b38fd2ad99e7bc23e4c3)`

### Issue
Cannot enable docker privileged containers, or raw_exec using JSON configuration. Based on the example HCL configs, AIUI I should be able to do the following:

`/etc/nomad.d/plugins.json`:

```
{
    "plugin": {
        "docker": {
            "config": {
                "allow_privileged": true
            }
        },
        "raw_exec": {
            "config": {
                "enabled": true
            }
        }
    }
}
```

Based on the `configSpec` defined in both plugins this looks correct, but errors on start:

```
 error parsing 'config': plugin-> invalid keys in plugin config 1: 1 error(s) occurred
  * invalid key: allow_privileged 
```

Removing the contents of the config block (leaving it an empty object) and I get `error parsing 'config': plugin-> plugin config 1 doesn't incude a name key` instead.

I've also tried making `plugin` a list of objects, and various other formats, and none of them work - I'm lost.

Is this supposed to work? If so, what is the correct JSON format? I can't find any tests which use a JSON config to set those settings so there's nothing to go off here except trying to read the code.

We currently use JSON configs with a 0.8.7 cluster and I'm trying to get ahead of the curve prepping our deployment setup to handle 0.9 properly.

### PR DESCRIPTION
**Pull request by preetapan**
This fixes a bug with JSON agent configuration parsing where the AST
for the plugin stanza had unnecessary flattening originating from hcl parsing. 

The workaround fixes the AST by popping off the flattened element and wrapping
it in a list. The workaround comes from [similar code in terraform](https://github.com/hashicorp/terraform-config-inspect/blob/8022a2663a706bf2f36bfaf7968a9234f9e5a571/tfconfig/load_legacy.go#L65)

There were no existing test cases for json parsing so I added a few.

fixes #5484

### CODE REVIEW

File: command/agent/config-test-fixtures/basic.json [hunk 1]
```diff
@@ -0,0 +1,288 @@
+{
```
Comment (notnoop) [inline command/agent/config-test-fixtures/basic.json]: Also maybe document how you generated this file?  This is neat.

File: command/agent/config_parse.go [hunk 2]
```diff
@@ -1014,6 +1014,11 @@ func parsePlugins(result *[]*config.PluginConfig, list *ast.ObjectList) error {
 		// Get the current plugin object
 		listVal := list.Items[i]
 
+		// Deal with json->hcl AST parsing incorrectness when directly nested
+		// items show up as additional keys. This currently only affects plugin
+		// configuration because args is not necessary. All other fields in the config
+		// have multiple keys and parse from json into the AST correctly.
```
Comment (notnoop) [inline command/agent/config_parse.go]: The comment is right in that we only want to target this one - but I don't agree with the rationale.

While it's true that other config stanzas require multiple entries, a user may still set one field only triggering this condition and without `unwrapLegacyHCLObjectKeysFromJSON` call, the error message might not be comprehensible.  However, given this is only relevant for nested entries, I don't see it happening elsewhere.

Beyond 0.9, we want to increase our test coverage.  I noticed that other `parse*` functions have `list = list.Elem()` call that's missing here, correctly so but not sure why.  Also, I noticed that it's easy to get parser to crash with simple config files, like:

```
# basic client stanza
{"client": {"server_join": {"retry_max": 3}}}

# sentinel path that seems similar to this one
{"sentinel": {
		  "import": {"custom-plugin": {"path": "/custom-path"}}
}}
```

In both of these cases, `list = list.Elem()` causes us to get an empty list that causes a crash when indexed.  The pattern causing them affect many stanzes here.

Comment (preetapan) [reply command/agent/config_parse.go]: agreed, there are more lingering bugs in json parsing. Found this open ticket from 2016 :( https://github.com/hashicorp/nomad/issues/1290. Let's track this for the next minor release, +1 for more test cases

File: command/agent/config-test-fixtures/basic.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (preetapan) [reply command/agent/config-test-fixtures/basic.json]: Done, noting that we use both testdata/ and other different names for test fixtures, should be made consistent in a future PR

File: command/agent/config_parse.go (hunk 2 above)
Comment (preetapan) [reply command/agent/config_parse.go]: Fixed the wording of the comment

**Pull request merged**